src/libstd/sync/mpsc/sync.rs

```diff
@@ -126,6 +127,28 @@ fn wait<'a, 'b, T>(lock: &'a Mutex<State<T>>,
     lock.lock().unwrap() // relock
 }
 
+/// Same as wait, but waiting at most `timeout`.
+fn wait_timeout_receiver<'a, 'b, T>(lock: &'a Mutex<State<T>>,
+                                    timeout: Duration,
+                                    mut guard: MutexGuard<'b, State<T>>)
+                                    -> MutexGuard<'a, State<T>>
+{
+    let (wait_token, signal_token) = blocking::tokens();
+    match mem::replace(&mut guard.blocker, BlockedReceiver(signal_token)) {
+        NoneBlocked => {}
+        _ => unreachable!(),
+    }
+    drop(guard);         // unlock
+    let timed_out = !wait_token.wait_timeout(timeout);   // block
```

Comment:
After working a bit more with the code, I'm pretty sure there can't be any races resetting the state if this wakes up normally (the senders fill it with `BlockedSender` or `NoneBlocked` depending on the capacity of the channel, so we don't need to reset any state).